src/librustc_mir/transform/mod.rs

```diff
@@ -98,7 +98,12 @@ fn mir_keys(tcx: TyCtxt<'_>, krate: CrateNum) -> &DefIdSet {
 }
 
 fn mir_built(tcx: TyCtxt<'_>, def_id: DefId) -> &Steal<BodyCache<'_>> {
-    let mir = build::mir_build(tcx, def_id);
+    let mut mir = build::mir_build(tcx, def_id);
+    if let ty::Generator(..) = tcx.type_of(def_id).kind {
```

Comment:
This should be is own query that calls `mir_built` so that we don't require optimisations to be run to get this value. I also think that using `mir_const` makes more sense here because dataflows already don't consider unreachable code.

Reply:
The problem is that `mir_validated` may be called (from `fn mir_borrowck`) before we process the generator auto trait obligations.

I could process the delayed generator obligations earlier, during `mir_borrowck` instead of `mir_typeck` (even though it's not really borrow checking).

> I also think that using mir_const makes more sense here because dataflows already don't consider unreachable code.

I'm not extremely clear on what `mir_const` does. Do we need to worry about changes to it causing this liveness computation to become dependent on optimizations?

Reply:
`mir_const` is the generated MIR with unreachable CFG nodes removed and goto chains simplified. To ensure that the query is run before `mir_const` is stolen, you can call the query in `mir_validated` (the same as for `mir_const_qualif`).

Reply:
Updated

Reply:
Updated

Reply:
This still isn't quite what I meant. I've left a couple of comments explaining this.